client/consensus/grandpa/finality_proof_test.go

```diff
@@ -0,0 +1,506 @@
+	authoritySetChanges.append(0, 5)
+	authoritySetChanges.append(1, 8)
+
+	proofOf6, err := proveFinality[
+		*BackendMock[string, uint, testHeader[string, uint],
+			*BlockchainBackendMock[string, uint, testHeader[string, uint]]],
+		string,
+		uint,
+		string,
+		dummyAuthID,
+		testHeader[string, uint],
+		*BlockchainBackendMock[string, uint, testHeader[string, uint]],
+	](
+		mockBackend,
+		authoritySetChanges,
+		6,
+		true,
+	)
+	require.NoError(t, err)
+
+	unknownHeaders := []testHeader[string, uint]{header7, header8}
+	expFinalityProof := &FinalityProof[string, uint, testHeader[string, uint]]{
+		Block:          "hash8",
+		Justification:  encJust,
+		UnknownHeaders: unknownHeaders,
+	}
+	require.Equal(t, expFinalityProof, proofOf6)
+
+	mockBlockchain2 := NewBlockchainBackendMock[string, uint, testHeader[string, uint]](t)
+	mockBlockchain2.EXPECT().Info().Return(dummyInfo).Once()
+	mockBlockchain2.EXPECT().ExpectBlockHashFromID(uint(8)).Return("hash8", nil).Times(2)
+	mockBlockchain2.EXPECT().Justifications("hash8").Return(&justifications, nil).Times(1)
+
+	mockBackend2 := NewBackendMock[string, uint, testHeader[string, uint],
+		*BlockchainBackendMock[string, uint, testHeader[string, uint]]](t)
+	mockBackend2.EXPECT().Blockchain().Return(mockBlockchain2).Times(4)
+
+	proofOf6WithoutUnknown, err := proveFinality[
+		*BackendMock[string, uint, testHeader[string, uint],
+			*BlockchainBackendMock[string, uint, testHeader[string, uint]]],
+		string,
+		uint,
+		string,
+		dummyAuthID,
+		testHeader[string, uint],
+		*BlockchainBackendMock[string, uint, testHeader[string, uint]],
+	](
+		mockBackend2,
+		authoritySetChanges,
+		6,
+		false,
+	)
+	require.NoError(t, err)
+
+	expFinalityProof = &FinalityProof[string, uint, testHeader[string, uint]]{
+		Block:         "hash8",
+		Justification: encJust,
+	}
+	require.Equal(t, expFinalityProof, proofOf6WithoutUnknown)
+}
+
+func TestFinalityProof_InLastSetFailsWithoutLatest(t *testing.T) {
+	dummyInfo := Info[uint]{
+		FinalizedNumber: 8,
+	}
+	mockBlockchain := NewBlockchainBackendMock[string, uint, testHeader[string, uint]](t)
+	mockBlockchain.EXPECT().Info().Return(dummyInfo).Once()
+
+	mockBackend := NewBackendMock[string, uint, testHeader[string, uint],
+		*BlockchainBackendMock[string, uint, testHeader[string, uint]]](t)
+	mockBackend.EXPECT().Blockchain().Return(mockBlockchain).Times(1)
+	mockBackend.EXPECT().Get(Key("grandpa_best_justification")).Return(nil, nil).Times(1)
+
+	// No recent authority set change, so we are in the authoritySetChangeIDLatest set, and we will try to pickup
+	// the best stored justification, for which there is none in this case.
+	authoritySetChanges := AuthoritySetChanges[uint]{}
+	authoritySetChanges.append(0, 5)
+
+	/*
+		TODO For reviewer: When justification is not stored in db, substrate returns None(), but we return error.
+		I think I like error, and we can handle accordingly, but will note this for reviewer feedback
```

Comment:
I don't think we should deviate.  Using a pointer and returning `nil` is valid IMO.  I don't think it's an error if a db is missing a value.